### DISCUSSION TITLE
web search doesn't work behind http proxy - proxy settings ignored

### DISCUSSION THREAD

**Discussion by ips972** [open-webui/open-webui#7008] [Issues]
Bug Report: HTTP Proxy Ignored for Web Scraping in Open WebUI (SearXNG Integration)
Description
When using SearXNG with Open WebUI, the initial connection to SearXNG successfully retrieves JSON search results. However, the subsequent web scraping process that fetches content from URLs within the results does not respect the configured HTTP proxy settings. This causes failures in environments where direct internet access is blocked, and a proxy is mandatory for outbound connections.

Despite attempting to configure HTTP proxy settings in various ways (e.g., environment variables, Docker options, and application-level configurations), the scraper fails to use the proxy. Upon inspecting the relevant Python file responsible for downloading URLs, it appears that proxy support is either not implemented or incorrectly handled.

Steps to Reproduce
Deploy Open WebUI with SearXNG integration enabled.
both using http_proxy correctly - open webui able to download models from HF, and seaxNG able to search the web. 
searxNG proxy setting given in it config file yaml  and works fine - either api call and manual user use work great.
enabled json response in settings. this also works fine. 
 here is the tcpdump (parcial) result from searxNG to open webui: question asked : whats new in ai? 

HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 27329
Server-Timing: total;dur=1622.189, render;dur=0, total_0_google;dur=697.018, total_1_wikipedia;dur=763.253, total_2_brave;dur=912.054, total_3_duckduckgo;dur=1026.366, total_4_qwant;dur=1347.023, total_5_wikidata;dur=1599.436, load_0_google;dur=643.497, load_1_wikipedia;dur=747.168, load_2_brave;dur=837.748, load_3_duckduckgo;dur=1006.767, load_4_qwant;dur=1300.943, load_5_wikidata;dur=1591.216
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Robots-Tag: noindex, nofollow
Referrer-Policy: no-referrer
Connection: close

{"query": "latest advancements in artificial intelligence 2024", "number_of_results": 0, "results": [{"url": "https://www.technologyreview.com/2024/01/04/1086046/whats-next-for-ai-in-2024/", "title": "What's next for AI in 2024", "content": "January 8, 2024 - In 2024, generative AI might actually become useful for the regular, non-tech person, and we are going to see more people tinkering with a million little AI models. State-of-the-art AI models, such as GPT-4 and Gemini, are multimodal, meaning they can process not only text but images and even ...", "thumbnail": "", "engine": "google", "parsed_url": ["https", "www.technologyreview.com", "/2024/01/04/1086046/whats-next-for-ai-in-2024/", "", "", ""], "template": "default.html", "engines": ["qwant", "google", "brave"], "positions": [7, 2, 1], "publishedDate": "2024-01-08T00:00:00", "score": 4.928571428571429, "category": "general"}, {"url": "https://www.techtarget.com/searchenterpriseai/tip/9-top-AI-and-machine-learning-trends", "title": "10 top AI and machine learning...........

Configure HTTP proxy settings:
Add http_proxy and https_proxy environment variables in the system.
Pass the proxy configuration via Docker:
bash
Copy code
docker run -d -p 3000:8080 --gpus all \
  -e http_proxy=http://proxy.example.com:3128 \
  -e https_proxy=http://proxy.example.com:3128 \
  --name open-webui \
  ghcr.io/open-webui/open-webui:cuda
Include proxy settings in the Open WebUI configuration (if applicable).
Perform a search in Open WebUI using SearXNG.
Observe that:
SearXNG returns the JSON response with search results.
The subsequent web scraping fails to retrieve content from the URLs due to lack of proxy usage.
Expected Behavior
The scraper should respect HTTP proxy settings and use the specified proxy for all outbound connections, including fetching content from the URLs returned in the SearXNG JSON results.

**Actual Behavior**
The scraper attempts direct internet connections, bypassing the configured HTTP proxy. As a result, web scraping fails in environments with restricted direct internet access. since no DNS resolve in the network is available. only http proxy browsing is allowed.

**Additional Information**
Configuration Attempts:

Tried setting http_proxy and https_proxy as environment variables.
Passed proxy settings during Docker container creation.
Manually modified the scraper's Python code to hardcode proxy settings.

**Code Inspection:**

The Python file responsible for downloading URLs appears not to include proxy support.
Observed the absence of proxy-related options in the code paths leading to URL fetching.
Environment Details:

**Open WebUI version:** v0.3.35  (main or cuda - same results)
**SearXNG integration**: Enabled
**Deployment method:** Docker
**Proxy type:** HTTP

**Proposed Solution**
Implement HTTP proxy support in the web scraping module.
Ensure the proxy settings are inherited from environment variables or explicitly passed through configurations.
Provide documentation on how to configure proxies for the scraping service.

**Impact**
This bug limits the functionality of Open WebUI in secure environments, preventing users from leveraging SearXNG effectively. Addressing this issue is critical for use cases where proxies are a standard requirement.

**debug logs:** 
ERROR [open_webui.apps.retrieval.main] [Errno -3] Temporary failure in name resolution
Traceback (most recent call last):
  File "/app/backend/open_webui/apps/retrieval/main.py", line 1165, in process_web_search
    loader = get_web_loader(urls)
             ^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/apps/retrieval/web/utils.py", line 90, in get_web_loader
    if not validate_url(url):
           ^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/apps/retrieval/web/utils.py", line 41, in validate_url
    return all(validate_url(u) for u in url)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/apps/retrieval/web/utils.py", line 41, in <genexpr>
    return all(validate_url(u) for u in url)
               ^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/apps/retrieval/web/utils.py", line 30, in validate_url
    ipv4_addresses, ipv6_addresses = resolve_hostname(parsed_url.hostname)
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/apps/retrieval/web/utils.py", line 48, in resolve_hostname
    addr_info = socket.getaddrinfo(hostname, None)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/socket.py", line 974, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
socket.gaierror: [Errno -3] Temporary failure in name resolution

**Comment by ips972** [comment]
may i ask why was this moved to discussions? it a bug in the code... its an issue not some this to chat about.

**Comment by gamesgao** [comment]
Aha, I think this is exactly what I need and seems this is not a bug recently and it have been there for several versions. 

I'm currently use Tools in OpenWebUI as a way to replace Web Search function inside OpenWebUI

**Reply by ips972** [reply]
> I'm currently use Tools in OpenWebUI as a way to replace Web Search function inside OpenWebUI

Hi, Yes this is a problem in  a closed network with an http proxy. 
can you PM me or paste it here, the tools solution your using, and maybe a little guidence on how to implement the solution?

**Reply by gamesgao** [reply]
Check the official page and the list of tool in that page. 
Or use this link https://openwebui.com/t/constliakos/web_search/
Which is from the #1 in the tool list

**Comment by ips972** [comment]
anyone? any ideas?

**Comment by codingstube** [comment]
Still no solution. Generating the web search request is working behind proxy. Also the search suggestions while typing is working. But if open-webui get the answers from searxng the following warning and timeouts appears:

```
Fetching pages:  33%|###3      | 1/3 [01:37<03:15, 97.90s/it]WARNI [langchain_community.document_loaders.web_base] Error fetching https://www.komoot.com/guide/6003/hiking-around-wachtendonk with attempt 1/3: Connection timeout to host https://www.komoot.com/guide/6003/hiking-around-wachtendonk. Retrying...
WARNI [langchain_community.document_loaders.web_base] Error fetching https://www.komoot.com/de-de/guide/6003/wandern-rund-um-wachtendonk with attempt 1/3: Connection timeout to host https://www.komoot.com/de-de/guide/6003/wandern-rund-um-wachtendonk. Retrying...
```

I also tried to add proxy configuration to the web loader but no luck:
```
  loader = get_web_loader(
            urls,
            proxies=proxies,
            verify_ssl=app.state.config.ENABLE_RAG_WEB_LOADER_SSL_VERIFICATION,
            requests_per_second=app.state.config.RAG_WEB_SEARCH_CONCURRENT_REQUESTS,
            timeout=15  # Increased timeout
        )
```

Final error:
```
ERROR [open_webui.apps.retrieval.main] The content provided is empty. Please ensure that there is text or data present before proceeding.
Traceback (most recent call last):
  File "/app/backend/open_webui/apps/retrieval/main.py", line 1344, in process_web_search
    save_docs_to_vector_db(docs, collection_name, overwrite=True)
  File "/app/backend/open_webui/apps/retrieval/main.py", line 799, in save_docs_to_vector_db
    raise ValueError(ERROR_MESSAGES.EMPTY_CONTENT)
ValueError: The content provided is empty. Please ensure that there is text or data present before proceeding.
```

**Reply by ips972** [reply]
well? have we had any progress with this?
its realy causing user in my network to suffer from this.. their all moving to chatGPT instead of my local models since its not really updated.

**Comment by ips972** [comment]
Just a quick update: all community tools appear to function through the HTTP proxy server, which is great! However, there's one significant issue: they sometimes work and sometimes don't. We have no idea what triggers this inconsistency, but it makes them unreliable for users who need a stable solution. We would greatly appreciate a fix to the built-in web browsing feature to support HTTP proxies, similar to the one added for YouTube links.

Tnx!

**Comment by ips972** [comment]
sorry for being a nag..  have we had any progress with this?

un fortunately , i have no coding experience and cannot add any in sight on solving this, chatGPT failed to update the utils.py .. so its up to you geniuses to solve it for us :-)

**Comment by zanxiong** [comment]
i think this is the fix for this:
https://github.com/open-webui/open-webui/commit/74c8690cd1514d004e70a8f8cbd168b4b930a6c2
https://github.com/open-webui/open-webui/commit/af10aba805119c58a02b228d72a9fb2923f77387
just make sure RAG_WEB_SEARCH_TRUST_ENV is true in ENV after v0.5.14.

**Reply by ips972** [reply]
didnt help. 

the python code that does the scraping from the seaxng json results. tries to do direct connection even though http_proxy and https_proxy are set.. it uses direc dns to resolve the results. and since only proxy connection works without local dns. every tjing fails. 

so the issus is that utils.py was initially built with no proxy suppot. ... at least this is what chatgpt says. the fix needs more modules to be imported for proxy support.. but they dont exist in the image. so the fix is needed by developer.

**Comment by menghui1206** [comment]
good solution

**Reply by ips972** [reply]
not for me.. look up. thats my situation.

**Comment by ips972** [comment]
just clearify, when pressing the check updates button. it works through proxy.. proving the other modules in the system are well built for proxy support.. this web scraping python file isnt..

**Comment by maidou-00** [comment]
I've noticed the same problem. I deployed SearXNG (as a docker and proxied), setup SearXNG as the search engine in OpenWebUI, also setup proxy for OpenWebUI. I use Litellm to invoke deepseek as deepseek is not natively supported by OpenwebUI. 
**Problem**: with OpenwebUI proxied(I setup http_proxy, https_proxy and no proxy, and make sure Litellm is in the no proxy list), it cannot call model from Litellm (no models were shown in OpenwebUI). When removed http_proxy, OpenWebUI can retrieve models from Litellm, but the search does not work.

When removing all the proxy settings for OpenWebUI, SearXNG can send searched website but OpenwebUI always tries to access the URLs...without proxy it won't be able to access censored website such as twitter so it just kept retrying and timeout.

**Comment by ips972** [comment]
I've come to realize that the built in web search is the most important option even though there are addons that do work. 
only the internal built in one has the "web search" button that users understand how to use.

we've been talking about this issue for a long time now , and in the same window for "web search" there has been a proxy setting added for youtube links.. it seems that we are only a few baby steps away from giving the proper python file the ability to pickup on http proxy settings in environment variables or a settings box in the ui. 

we are currently in version 0.6.9  and i've opened this issue since version 0.3.35 
this problem is relevent to all web search services' since they all bring back a list of web sites in json for the utils.py to scrape - and this fails for all no just searxng.

**Reply by ananthb** [reply]
I'm in the exact same boat. Not sure what eles to add but this needs to be fixed.

**Comment by ukpark** [comment]
I had the same problem few month ago. Today I succeed to run the web search behind the proxy. 

1. I tried to set  Enviroment Variable (WEB_SEARCH_TRUST_ENV=True) when I run the webui docker. But this is not working. 
2. After I started the webui, I go to Admin Panel - Setting - Web Search, Then enable  "Trust Proxy Environment". This is working.

I used the google_pse search for the web search.

**Reply by Classic298** [reply]
Let me update the docs to reflect that

**Reply by Classic298** [reply]
https://github.com/open-webui/docs/pull/938

**Reply by ASRSI** [reply]
> I had the same problem few month ago. Today I succeed to run the web search behind the proxy.
> 
>     1. I tried to set  Enviroment Variable (WEB_SEARCH_TRUST_ENV=True) when I run the webui docker. But this is not working.
> 
>     2. After I started the webui, I go to Admin Panel - Setting - Web Search, Then enable  "Trust Proxy Environment". This is working.
> 
> 
> I used the google_pse search for the web search.

I have the exact same problem (no web search with correct proxy env settings) and "Trust Proxy Environment" is enabled, but it still is nor working.
